superduperdb/base/datalayer.py

```diff
@@ -857,55 +857,68 @@ def _add(
-        already_exists = (
-            isinstance(object.version, int) and object.version in existing_versions
-        )
-
-        if already_exists:
-            return self._update_component(
-                object, dependencies=dependencies, parent=parent
+        try:
+            jobs = list(dependencies)
+            object.db = self
+            object.pre_create(self)
+            assert hasattr(object, 'identifier')
+            assert hasattr(object, 'version')
+
+            if not isinstance(object.identifier, variables.Variable):
+                existing_versions = self.show(object.type_id, object.identifier)
+            else:
+                existing_versions = []
+            already_exists = (
+                isinstance(object.version, int) and object.version in existing_versions
             )
 
-        if object.version is None:
-            if existing_versions:
-                object.version = max(existing_versions) + 1
-            else:
-                object.version = 0
+            if already_exists:
+                return self._update_component(
+                    object, dependencies=dependencies, parent=parent
+                )
+
+            if object.version is None:
+                if existing_versions:
+                    object.version = max(existing_versions) + 1
+                else:
+                    object.version = 0
 
-        serialized = object.dict().encode(leaves_to_keep=(Component,))
+            serialized = object.dict().encode(leaves_to_keep=(Component,))
 
-        children = [
-            v for k, v in serialized['_leaves'].items() if isinstance(v, Component)
-        ]
+            children = [
+                v for k, v in serialized['_leaves'].items() if isinstance(v, Component)
+            ]
 
-        jobs.extend(self._add_child_components(children, parent=object))
+            jobs.extend(self._add_child_components(children, parent=object))
 
-        for k, v in serialized['_leaves'].items():
-            if isinstance(v, Component):
-                serialized['_leaves'][k] = f'%{v.id}'
+            for k, v in serialized['_leaves'].items():
+                if isinstance(v, Component):
+                    serialized['_leaves'][k] = f'%{v.id}'
 
-        serialized = self.artifact_store.save_artifact(serialized)
+            serialized = self.artifact_store.save_artifact(serialized)
 
-        self.metadata.create_component(serialized)
+            self.metadata.create_component(serialized)
 
-        if parent is not None:
-            self.metadata.create_parent_child(parent, object.uuid)
+            if parent is not None:
+                self.metadata.create_parent_child(parent, object.uuid)
+
+            object.post_create(self)
+            self._add_component_to_cache(object)
+            these_jobs = object.schedule_jobs(self, dependencies=dependencies)
+            jobs.extend(these_jobs)
+            return jobs
+        except Exception:
```

Comment:
Generally not good practice to use a bare exception. Better would be to handle specific types of `Exception` in a piecemeal way, giving informative error messages.